ddl/check.lua

```diff
@@ -628,75 +628,113 @@ local function check_index(i, index, space)
-                space.name, 'bucket_id'
+                space_name, 'bucket_id'
             )
         end
 
-        if #space.indexes['bucket_id'].parts ~= 1 then
+        if #bucket_id_index.parts ~= 1 then
             return nil, string.format(
                 "spaces[%q].indexes[%q].parts: bucket_id index can't be composite (1 part expected, got %d parts)",
-                space.name, 'bucket_id', #space.indexes['bucket_id'].parts
+                space_name, 'bucket_id', #bucket_id_index.parts
             )
         end
 
-        if space.indexes['bucket_id'].parts[1].path ~= 'bucket_id' then
+        if bucket_id_index.parts[1].path ~= 'bucket_id' then
             return nil, string.format(
                 "spaces[%q].indexes[%q].parts[1].path: invalid field reference " ..
                 "(reference to bucket_id expected, got %s)",
-                space.name, 'bucket_id', space.indexes['bucket_id'].parts[1].path
+                space_name, 'bucket_id', bucket_id_index.parts[1].path
             )
         end
     end
 
+    return true
+end
+
+local function check_sharding_func(space)
+    if not space.sharding_func then
+        return true
+    end
+
+    do -- check that sharding_func format is valid
+        if type(space.sharding_func) ~= 'string' and type(space.sharding_func) ~= 'table' then
+            return nil, string.format(
+                "spaces[%q].sharding_func: bad value" ..
+                " (string or table expected, got %s)",
+                space.name, type(space.sharding_func)
+            )
+        end
+
+        if type(space.sharding_func) == 'string' then
+            local known_sharding_funcs = {
+                ['mpcrc32'] = true,
+                ['strcrc32'] = true,
+                ['vshard.router.bucket_id_strcrc32()'] = true,
+                ['vshard.router.bucket_id_mpcrc32()'] = true,
+            }
+            if not known_sharding_funcs[space.sharding_func] and rawget(_G, space.sharding_func) == nil then
+                return nil, string.format(
+                    "spaces[%q].sharding_func: unknown sharding_func %q",
+                    space.name, space.sharding_func
+                )
+            end
+        else
+            if not space.sharding_func.body then
+                return nil, string.format(
+                    "spaces[%q].sharding_func: sharding_func exists in the space as table," ..
+                    "but there's no 'body' field in sharding_func section",
+                    space.name
+                )
+            end
+
+            if type(space.sharding_func.body) ~= 'string' then
+                return nil, string.format(
+                    "spaces[%q].sharding_func.body: bad value" ..
+                    " (string expected, got %s)",
+                    space.name, type(space.sharding_func.body)
+                )
+            end
+        end
+    end
+
+    return true
+end
+
+local function check_sharding_key(space)
```

Comment:
NB: Now the function does nothing when there is no `space.sharding_key`. Before the change it did check that there is no `bucket_id` space and index. I understand the change: it is to don't do the same check twice.

Tehcnically this function is available for a user, but the ddl submodules are not mentioned in the README, so it is the grey zone. I would consider them all as internal (so the change is okay, at least from the compatibility side of view).

Anyway, commented it here to don't forget to look again.

Reply:
I change `check_sharding_key` and `check_sharding_func` functions visibility. Now both of them are internal:

```
     check_index = check_index,
-    check_field = check_field,
-    check_sharding_key = check_sharding_key,
-    check_sharding_func = check_sharding_func
+    check_field = check_field
 }
```

Should I change other check functions (`check_index` etc.) to internal too?

Reply:
Okay. No, I don't think we should revisit functions visibility at whole within this task.